clients/src/main/java/org/apache/kafka/clients/producer/KafkaProducer.java

```diff
@@ -920,23 +923,28 @@ private void throwIfProducerClosed() {
                         " to class " + producerConfig.getClass(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG).getName() +
                         " specified in key.serializer", cce);
             }
-            byte[] serializedValue;
-            try {
-                serializedValue = valueSerializer.serialize(record.topic(), record.headers(), record.value());
-            } catch (ClassCastException cce) {
-                throw new SerializationException("Can't convert value of class " + record.value().getClass().getName() +
+
+            setReadOnly(record.headers());
```

Comment:
I think this line is what is causing the test to fail. It sets the header as read only, but in the test, we try to write to it.
From the stack trace,

```
at org.apache.kafka.common.header.internals.RecordHeaders.canWrite(RecordHeaders.java:122)
	at org.apache.kafka.common.header.internals.RecordHeaders.add(RecordHeaders.java:61)
	at org.apache.kafka.common.header.internals.RecordHeaders.add(RecordHeaders.java:68)
	at kafka.api.PlaintextConsumerTest$SerializerImpl.serialize(PlaintextConsumerTest.scala:76)
	at kafka.api.PlaintextConsumerTest$SerializerImpl.serialize$(PlaintextConsumerTest.scala:75)
	at kafka.api.PlaintextConsumerTest$$anon$1.serialize(PlaintextConsumerTest.scala:142)
	at kafka.api.PlaintextConsumerTest$$anon$1.serialize(PlaintextConsumerTest.scala:142)
	at org.apache.kafka.clients.producer.KafkaProducer.doSend(KafkaProducer.java:937)```

Reply:
fixed this one
sorry for making you look for the cause, please have a look at the PR again